Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,463 @@
+            _npcFaction.AddFaction(ev.Target, "Revolutionary");
+            EnsureComp<RevolutionaryComponent>(ev.Target);
+            _sharedStun.TryParalyze(ev.Target, stunTime, true);
+            if (ev.User != null)
+            {
+                _adminLogManager.Add(LogType.Mind, LogImpact.Medium, $"{ToPrettyString(ev.User.Value)} converted {ToPrettyString(ev.Target)} into a Revolutionary");
+            }
+
+            var mind = _mindSystem.GetMind(ev.Target);
+            if (mind != null && mind.OwnedEntity != null)
+            {
+                _mindSystem.AddRole(mind, new RevolutionaryRole(mind, _prototypeManager.Index<AntagPrototype>("Rev")));
+                if (mind.Session != null)
+                {
+                    var message = Loc.GetString("rev-role-greeting");
+                    var wrappedMessage = Loc.GetString("chat-manager-server-wrap-message", ("message", message));
+                    _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, message, wrappedMessage, default, false, mind.Session.ConnectedClient, Color.Red);
+                }
+            }
+        }
+    }
+
+    public void OnHeadRevAdmin(Mind.Mind mind, IPlayerSession headRev)
+    {
+        var revRule = EntityQuery<RevolutionaryRuleComponent>().FirstOrDefault();
+        if (revRule == null)
+        {
+            GameTicker.StartGameRule("Revolutionary");
+        }
+        if (mind.OwnedEntity != null)
+        {
+            EnsureComp<RevolutionaryRuleComponent>(mind.OwnedEntity.Value);
+            GiveHeadRevRole(mind, headRev);
+            AssignCommandStaff();
+        }
+    }
+    private void OnMobStateChanged(EntityUid uid, RevolutionaryRuleComponent comp, MobStateChangedEvent ev)
+    {
+        if (ev.NewMobState == MobState.Dead || ev.NewMobState == MobState.Invalid)
+            CheckFinish();
+    }
+    /// <summary>
+    /// Checks if all Head Revs are dead and if all command is dead to either end the round or remove all revs. Or both.
+    /// </summary>
+    private void CheckFinish()
+    {
+        var stunTime = TimeSpan.FromSeconds(4);
+        var headJobs = _prototypeManager.Index<DepartmentPrototype>("Command");
+        var secJobs = _prototypeManager.Index<DepartmentPrototype>("Security");
+
+        var query = AllEntityQuery<RevolutionaryRuleComponent, GameRuleComponent>();
+        while (query.MoveNext(out var uid, out var revs, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleAdded(uid, gameRule))
+                continue;
+            var headRevs = EntityQuery<HeadRevolutionaryComponent, MobStateComponent>(true);
+            var inRound = 0;
+            var dead = 0;
+            foreach (var rev in headRevs)
+            {
+                if (rev.Item2.CurrentState == MobState.Dead || rev.Item2.CurrentState == MobState.Invalid)
+                {
+                    dead++;
+                }
+                inRound++;
+            }
+
```

Comment:
Use entityqueryenumerators as mentioned above.

Reply:
This breaks the for each loop so any suggestions?